File: sorting/quicksort.c
#include <stdio.h>

void quicksort(int number, int* array) {
    if (number <= 1)
        return;
    int pivot = array[0];
    int i = 1;
    int j = number-1;
    do {
        while (i < number && array[i] <= pivot) i++;
        while (array[j] > pivot) j--;
        if (i < j) {
            int temp = array[i];
            array[i] = array[j];
            array[j] = temp;
            i++;
            j--;
        }
    } while(i <= j);
    array[0] = array[j];
    array[j] = pivot;
    quicksort(j, array);
    quicksort(number-i, &array[i]);
}

int main(void) {
    int array[8] = { 25, 48, 37, 12, 57, 86, 33, 92 };
    quicksort(8, array);
    printf("sorted array: ");
    for (int i = 0; i < 8; i++)
        printf("%d ", array[i]);
    printf("n");
    return 0;
}

File: sorting/quicksort.ts
function swap_numbers(arr: number[], i: number, j: number) {
    [arr[i], arr[j]] = [arr[j], arr[i]];
}

function quicksort(array: number[], start: number, end: number) {
    if (start >= end) return;
    const pointer = rearrange(array, start, end);
    quicksort(array, start, pointer - 1);
    quicksort(array, pointer + 1, end);
}

function rearrange(array: number[], start: number, end: number): number {
    const pivot = array[end];
    let pointer = start;
    for (let i = start; i < end; i++) {
        if (array[i] < pivot) {
            swap_numbers(array, pointer, i);
            pointer++;
        }
    }
    swap_numbers(array, pointer, end);
    return pointer;
}

const values: Array<number> = [5, 3, 1, 2, 4];
quicksort(values, 0, values.length - 1);
console.log(values);
